modules/account-lib/src/coin/hbar/tokenTransferBuilder.ts

```diff
@@ -0,0 +1,123 @@
+      const tokenAmount = tokenTransferAmount[tokenName] || new BigNumber(0);
+      tokenTransferAmount[tokenName] = tokenAmount.plus(recipient.amount);
+      if (!(tokenName in tokenTransferData)) {
+        tokenTransferData[tokenName] = [
+          {
+            accountID: this.buildAccountData(this._source.address),
+            amount: Long.fromInt(0),
+          },
+        ];
+      }
+      tokenTransferData[tokenName].push(
+        { accountID: this.buildAccountData(recipient.address), amount: Long.fromString(recipient.amount) }, // recipient
+      );
+    });
+
+    Object.entries(tokenTransferData).forEach(([tokenName, tokenData]) => {
+      tokenData[0].amount = Long.fromString(tokenTransferAmount[tokenName].toString()).negate(); // update sender send amount
+      const tokenId = getHbarTokenFromTokenName(tokenName)!.tokenId!;
+      tokenTransfers.push({ token: this.buildTokenId(tokenId), transfers: tokenData });
+    });
+
+    return tokenTransfers;
+  }
+
+  private buildTokenId(tokenId: string): proto.TokenID {
+    const tokenData = TokenId.fromString(tokenId);
+    return new proto.TokenID({
+      tokenNum: tokenData.num,
+      realmNum: tokenData.realm,
+      shardNum: tokenData.shard,
+    });
+  }
+
+  /** @inheritdoc */
+  initBuilder(tx: Transaction): void {
+    super.initBuilder(tx);
+    const transferData = tx.txBody.cryptoTransfer!;
+    if (isTokenTransfer(transferData)) {
+      this.initTokenTransfers(transferData.tokenTransfers!);
+    }
+  }
+
+  /**
+   * Initialize the transfer specific data, getting the recipient account
+   * represented by the element with a positive amount on the transfer element.
+   * The negative amount represents the source account so it's ignored.
+   *
+   * @param {proto.IAccountAmount[]} transfers array of objects which contains accountID and transferred amount
+   */
+  protected initTokenTransfers(tokenTransfers: proto.ITokenTransferList[]): void {
+    tokenTransfers.map((tokenTransfer) => {
+      if (!tokenTransfer.token) {
+        throw new BuildTransactionError('Invalid transaction: missing token id');
+      }
+      if (!tokenTransfer.transfers) {
+        throw new BuildTransactionError('Invalid transaction: missing transfer data');
+      }
+
+      const token = getHbarTokenFromId(stringifyTokenId(tokenTransfer.token!), this._coinConfig.network);
+      tokenTransfer.transfers.forEach((transferData) => {
+        const amount = Long.fromValue(transferData.amount!);
+        if (amount.isPositive()) {
+          if (!token) {
+            throw new BuildTransactionError('Invalid transaction: invalid token id');
+          }
+          this.send({
+            address: stringifyAccountId(transferData.accountID!),
+            amount: amount.toString(),
+            tokenName: token.name,
+          });
+        }
+      });
+    });
+  }
+
+  // region Transfer fields
+  /** @inheritdoc */
+  send(recipient: Recipient): this {
```

Comment:
I like the ability to have multi recipient transactions, but I think we should at least restrict multiple tokens for sanity. We can add a check on `Recipient` for unique tokens here. Let me know what you think.

Reply:
Yeah after thinking about it, it's much complicated in some way like counting number of diff token for fee calculation and explainTransaction as the order messed up if parsed from TxHex, so I think skipping it for now will be better.